Add support for HD44780-compatible WINSTAR WEH00xxyyA OLED Displays

mskalski opened the pull request:
Added support for WINSTAR OLED displays - almost compatible to
HD44780 LCDs. These displays require slightly different initialization sequence,
especially required on display reinitialization without powering it off first.

Also added support for handling backlight using built-in command
for brightness setting.

Tested on Raspberry PI and WINSTAR WEH001604ALPP5N00001 OLED display
connected through I2C PCF8574 connection (4-bits obviously) and
WINSTAR WEH002004ALPP5N00001 OLED. Should work on any connection.

mskalski committed:
Add support for HD44780-compatible WINSTAR WEH00xxyyA OLED Displays

Added support for WINSTAR OLED displays - almost compatible to
HD44780 LCDs. These displays require slightly different initialization sequence,
especially required on display reinitialization without powering it off first.

Also added support for handling backlight using built-in command
for brightness setting.

Tested on Raspberry PI and WINSTAR WEH001604ALPP5N00001 OLED display
connected through I2C PCF8574 connection (4-bits obviously) and
WINSTAR WEH002004ALPP5N00001 OLED. Should work on any connection.

haraldg commented:
Looks almost perfect to me. The only thing missing is updating the LCDd.conf with the new (commented out) options you are introducing.

I'd have prefered the initialization part and the backlight part to be different commits, but never mind that now.

@trex2000 this PR seems to do something very similar to what might work for your PR #89 too. AFAICS only the register values are differnt. Maybe you can have a look and tell us, if this code can be generalized to cover your case as well? I'm thinking along the lines of renaming WinstarMapBl to something like InternalBl and introduce a new configuration setting for the actual register value.

mskalski commented:
Thanks for review

I thought about separating commits, but actually initialization depends on value of WinstarMapBl to set correct brightness.

Yes, I forgot about example in LCDd.conf, I'll fix it, thanks for pointing this out.

As you probably noted I slightly changed order of sending commands, which are sent to HDI44780 controller, so they may influence other compatible displays. I tested my change also on one connected through i2c backpack (I had no other hardware to test with) and it works well, but maybe someone else could test it also?

I thought a little bit more about my PR, and I think it can be done better:

1. Always on initialization turn off display (not only for Winstar OLED), to hide possibly displaying garbage. It should be change compatible to other displays.
2. Make WindstarMapBl independent of Backlight setting, to allow simultaneously handling internal brightness and external BL if someone needs it, but I feel it is a little overkill.

Any comments are welcome.😀

Regarding #89:
What do you mean setting register value?  In config file? If yes, I think it could be good idea for advanced users, but most (I think) users just know type of display and are not aware of commands required to handle internals. Feature I added was quite hidden, just found one note on some forum, I can't even remember where, during searching resolution for problems with reinitialization of WindstarMapBl OLED; In manual it's not documented, just mentioned.

So I think there should be options to turn on specific features (documented ie. in lcdproc manual) and they shall just work. If anyone would be interested in, he can look to source code how it is achieved.

haraldg commented:
re testing: I don't have a HD44780 so I can't help you there. You can ask on the ML or directly request feedback from previous commiters to the driver in this PR. I will wait with merging the PR if you think more testing is necessary.

re 1: Sounds reasonable

re 2: I think your code is fine as it is. But you know the display best and know what's useful and what is not.

re #89: Yes, I mean in config file. I'm not thinking about advanced users, but about reducing code duplication. For now we would have the register values commented out in the configuration file and the user would uncomment the section describing his display. In the long run I hope to extend the configuration language like outlined in https://github.com/lcdproc/lcdproc/issues/49#issuecomment-280653594 to make this even more comfortable.

mskalski committed:
Added commented WINSTAR-OLED options for reference

mskalski committed:
HD44780: always perform initialization in DISPLAY OFF state

Change initialization sequence for ALL HD44780 displays to
first turn off display, perfrom other configuration (font size change,
data width, font selection, ...) and then turn display on.

mskalski commented:
re #89: I think it is similar to this PR, but slightly different. _PT6314 VFD_ has four levels of brightness, _WINSTAR OLED_ has two. To do things right, these levels shall be mapped to configured `Brightness` and `OffBrightness` and I don't see a way to do it straight from configuration file. Especially it can  possibly be called at different stages, not only on backlight setting but also during initialization.

I think it is safer and clearer to add code to driver and one more option which enables this behavior (as I did with `WinstarMapBl`). My change was easier as it is caused by a change in (possibly incompatible) initialization sequence and requirements, so for this introduced `WinstarMode` option, so when former is enabled it means someone has this kind of screen and first is only as a side effect.  :smile:

Change for VFD can be by default disabled, as it is not handled by other display types and sometimes can be .

Or we can introduce something like `Subtype` or `SubModel` option, something similar to ideas in #49 or #79, which handle _Command Set differencies_ for HD44780 display(s). There are not so many now yet though: `ExtendedMode`, `WinstarMode` for now and new _VFD PT6314_.

On the other hand I think your proposition of introducing options for defining command(s) to on and off backlight is worth of implementing as last resort solution for future(maybe even unknown yet) displays.

trex2000 commented:
Thanks for joining these 2 commits. 
So, for the PT6314 driver, there are 2 bits in additionin FUNCSET register, that handle the VFD brightness:
00 - 100%
01 - 75%
02 - 50%
03 - 25%
So there are 4 different brightness levels.
In my commits, I've mapped these to the backlight setting rather than brightness, as usually (also XBMC plugin does this) the backlight (on or off) is set by client during runtime. This way, when there is an 'event' the backlight was turned on, and after a timeut, the backlight was turned off. For the VFD, this meant that in case of an 'event', the XBMC plugin requested 100% 
brightness on the VFD and after a while the 'backlight' was turned off, so the VFD's brigthness was set back to minimum value, 25%. This give a similar behavior like any commercial AV receiver has.
Now, the problem was that the FUCSET register was set once during init time, as it contained no information that was supposed to be set during runtime. But with brightness this changed, therefore I had to store the value of the FUNCSET register that was set at init, and use it also during runtime when the backlight request was changed. 
This is the modification that I've done.
Indeed, yours, for the OLED looks very similar, so I would also propose to somehow merge the implementation.

http://www.artronic.eu/pl/p/VFD-AC-1602H-BG-5V/937
This is the display I'm using and here's the datasheet:
https://www.newhavendisplay.com/app_notes/PT6314.pdf

trex2000 commented:
Yes, also to me, instead of 100% brightness, maybe it's safer to have it initialized with 25% only, therefore value 11. I wouldn't turn if off in this case, as by default nobody would set the parameter correctly, therefore the display won't be initialized. Also the 00 means 100% , therefore maximum brightness, and I think also the HW manufactureres thought it this way. In case there is no 'differentiated' support (so only generic instruction set for the HD44780 is available) the display would still work, but without brightness control.

mskalski commented:
I was thinking about above comments and these are my propositions:

Introduce `Model` option, as in other display types. It should be used to select command differences incompatible between various hd44780 clones. Here now would be `standard`, `winstar_oled`, `pt6314_vfd` or maybe more generic names. For now `pt6314_vfd` is only defined for handling internal backlight. Value `winstar_oled` causes slightly different initialization sequence and allows to choose internal backlight commands.

Modify existing `Backlight` option from Boolean to enumeration: `none`, `external`, `internal`,`internalCmds`and with currently existing Boolean values `yes`/`no` mapped to `external` and `none`.

* `internal` means that backlight is handled using internal commands according to selected model. Specific commands would be coded according to `Brightness` and `OffBrightness` options.

* `internalCmds` would mean that commands for turning on and off backlight are given with options `BacklightOnCmd` and `BacklightOffCmd`, which would be treated as catch up (last resort) for other types of displays which have similar features.

* `external` means same as for now `yes` - handling backlight using external pin.

I would rather not left only these `internalCmds` options. First they may interfere with standard initialization (for example for VFD one need to define interface bit lenght 4/8 bits additionally to specifying brightness). Second it's impossible to handle brightness set through driver interface this way - functions `get_brightness()` and `set_brightness()`.

It is tempting to allow combinations of some above options, to allow hypothetically possible situation of setting backlight  with both means (ie. bright display and extra led for glowing buttons) :smile:

I'll make new pull request with these proposed changes, with commits for enabling model selection and then backlight handling.

haraldg commented:
I have to admit that since there is no way to change the brightness from the client API, I completely forgot that we brightness control in the driver API. With that in mind I guess your approach makes sense.
